// appium-ios-device/build/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Xctest = exports.TESTMANAGERD_CHANNEL = exports.INSTRUMENT_CHANNEL = exports.services = exports.utilities = exports.Usbmux = void 0;
const usbmux_1 = __importDefault(require("./lib/usbmux"));
Object.defineProperty(exports, "Usbmux", { enumerable: true, get: function () { return usbmux_1.default; } });
const utilities = __importStar(require("./lib/utilities"));
exports.utilities = utilities;
const services = __importStar(require("./lib/services"));
exports.services = services;
var instrument_1 = require("./lib/instrument");
Object.defineProperty(exports, "INSTRUMENT_CHANNEL", { enumerable: true, get: function () { return instrument_1.INSTRUMENT_CHANNEL; } });
var testmanagerd_1 = require("./lib/testmanagerd");
Object.defineProperty(exports, "TESTMANAGERD_CHANNEL", { enumerable: true, get: function () { return testmanagerd_1.TESTMANAGERD_CHANNEL; } });
var xctest_1 = require("./lib/xctest");
Object.defineProperty(exports, "Xctest", { enumerable: true, get: function () { return xctest_1.Xctest; } });
exports.default = usbmux_1.default;
