// appium-xcuitest-driver/build/lib/commands/enum.js
"use strict";
/**
 * Mapping of permission resource name to identifier.
 * @see https://developer.apple.com/documentation/xctest/xcuiprotectedresource?language=objc
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.BatteryState = exports.AppState = exports.ThermalState = exports.AuthorizationStatus = exports.PermissionService = void 0;
var PermissionService;
(function (PermissionService) {
    PermissionService[PermissionService["calendar"] = 2] = "calendar";
    PermissionService[PermissionService["camera"] = 6] = "camera";
    PermissionService[PermissionService["contacts"] = 1] = "contacts";
    PermissionService[PermissionService["health"] = -1073741827] = "health";
    PermissionService[PermissionService["homekit"] = 8] = "homekit";
    PermissionService[PermissionService["keyboardnet"] = -1073741825] = "keyboardnet";
    PermissionService[PermissionService["location"] = -1073741826] = "location";
    PermissionService[PermissionService["medialibrary"] = 7] = "medialibrary";
    PermissionService[PermissionService["microphone"] = 5] = "microphone";
    PermissionService[PermissionService["photos"] = 4] = "photos";
    PermissionService[PermissionService["reminders"] = 3] = "reminders";
    PermissionService[PermissionService["systemroot"] = 1073741824] = "systemroot";
    PermissionService[PermissionService["userdesktop"] = 1073741825] = "userdesktop";
    PermissionService[PermissionService["userdocuments"] = 1073741827] = "userdocuments";
    PermissionService[PermissionService["userdownloads"] = 1073741826] = "userdownloads";
    PermissionService[PermissionService["bluetooth"] = -1073741824] = "bluetooth";
})(PermissionService || (exports.PermissionService = PermissionService = {}));
/**
 * @see https://developer.apple.com/documentation/corelocation/clauthorizationstatus
 */
var AuthorizationStatus;
(function (AuthorizationStatus) {
    AuthorizationStatus[AuthorizationStatus["notDetermined"] = 0] = "notDetermined";
    AuthorizationStatus[AuthorizationStatus["restricted"] = 1] = "restricted";
    AuthorizationStatus[AuthorizationStatus["denied"] = 2] = "denied";
    AuthorizationStatus[AuthorizationStatus["authorizedAlways"] = 3] = "authorizedAlways";
    AuthorizationStatus[AuthorizationStatus["authorizedWhenInUse"] = 4] = "authorizedWhenInUse";
})(AuthorizationStatus || (exports.AuthorizationStatus = AuthorizationStatus = {}));
/**
 * Thermal state of an iOS device.
 *
 * Returned (since iOS 11.0) within a {@linkcode DeviceInfo} response.
 * @see https://developer.apple.com/documentation/foundation/nsprocessinfothermalstate
 */
var ThermalState;
(function (ThermalState) {
    /**
     * The thermal state is within normal limits.
     */
    ThermalState[ThermalState["NSProcessInfoThermalStateNominal"] = 0] = "NSProcessInfoThermalStateNominal";
    /**
     * The thermal state is slightly elevated.
     */
    ThermalState[ThermalState["NSProcessInfoThermalStateFair"] = 1] = "NSProcessInfoThermalStateFair";
    /**
     * The thermal state is high.
     */
    ThermalState[ThermalState["NSProcessInfoThermalStateSerious"] = 2] = "NSProcessInfoThermalStateSerious";
    /**
     * The thermal state is significantly impacting the performance of the system and the device needs to cool down.
     */
    ThermalState[ThermalState["NSProcessInfoThermalStateCritical"] = 3] = "NSProcessInfoThermalStateCritical";
})(ThermalState || (exports.ThermalState = ThermalState = {}));
/**
 * Application state code.
 * @see https://developer.apple.com/documentation/xctest/xcuiapplicationstate?language=objc
 */
var AppState;
(function (AppState) {
    /**
     * The application's current state is not known.
     */
    AppState[AppState["XCUIApplicationStateUnknown"] = 0] = "XCUIApplicationStateUnknown";
    /**
     * The application is not running
     */
    AppState[AppState["XCUIApplicationStateNotRunning"] = 1] = "XCUIApplicationStateNotRunning";
    /**
     * The application is running in the background, but is suspended.
     */
    AppState[AppState["XCUIApplicationStateRunningBackgroundSuspended"] = 2] = "XCUIApplicationStateRunningBackgroundSuspended";
    /**
     * The application is running in the background.
     */
    AppState[AppState["XCUIApplicationStateRunningBackground"] = 3] = "XCUIApplicationStateRunningBackground";
    /**
     * The application is running in the foreground.
     */
    AppState[AppState["XCUIApplicationStateRunningForeground"] = 4] = "XCUIApplicationStateRunningForeground";
})(AppState || (exports.AppState = AppState = {}));
/**
 * Battery state
 * @see {@linkcode BatteryInfo}
 */
var BatteryState;
(function (BatteryState) {
    /**
     * The battery state is unknown.
     */
    BatteryState[BatteryState["UIDeviceBatteryStateUnknown"] = 0] = "UIDeviceBatteryStateUnknown";
    /**
     * The device is running on battery and discharging.
     */
    BatteryState[BatteryState["UIDeviceBatteryStateUnplugged"] = 1] = "UIDeviceBatteryStateUnplugged";
    /**
     * The device is plugged in, less than 100%, and charging.
     */
    BatteryState[BatteryState["UIDeviceBatteryStateCharging"] = 2] = "UIDeviceBatteryStateCharging";
    /**
     * The device is plugged in and is at 100% charge.
     */
    BatteryState[BatteryState["UIDeviceBatteryStateFull"] = 3] = "UIDeviceBatteryStateFull";
})(BatteryState || (exports.BatteryState = BatteryState = {}));
