// appium-remote-debugger/build/lib/mixins/property-accessors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAppDict = getAppDict;
exports.getAppIdKey = getAppIdKey;
exports.setAppIdKey = setAppIdKey;
exports.getRcpClient = getRcpClient;
exports.getPageIdKey = getPageIdKey;
exports.setPageIdKey = setPageIdKey;
exports.getIsSafari = getIsSafari;
exports.getIncludeSafari = getIncludeSafari;
exports.getBundleId = getBundleId;
exports.getAdditionalBundleIds = getAdditionalBundleIds;
exports.getSkippedApps = getSkippedApps;
exports.getClientEventListeners = getClientEventListeners;
exports.getPageLoading = getPageLoading;
exports.setPageLoading = setPageLoading;
exports.getGarbageCollectOnExecute = getGarbageCollectOnExecute;
exports.getNavigatingToPage = getNavigatingToPage;
exports.setNavigatingToPage = setNavigatingToPage;
exports.setCurrentState = setCurrentState;
exports.setConnectedDrivers = setConnectedDrivers;
exports.getPageLoadDelay = getPageLoadDelay;
exports.setPageLoadDelay = setPageLoadDelay;
exports.getPageLoadStartegy = getPageLoadStartegy;
exports.getPageReadyTimeout = getPageReadyTimeout;
function getAppDict(instance) {
    return instance['_appDict'];
}
function getAppIdKey(instance) {
    return instance['_appIdKey'];
}
function setAppIdKey(instance, value) {
    instance['_appIdKey'] = value;
}
function getRcpClient(instance) {
    return instance['_rpcClient'];
}
function getPageIdKey(instance) {
    return instance['_pageIdKey'];
}
function setPageIdKey(instance, value) {
    instance['_pageIdKey'] = value;
}
function getIsSafari(instance) {
    return instance['_isSafari'];
}
function getIncludeSafari(instance) {
    return instance['_includeSafari'];
}
function getBundleId(instance) {
    return instance['_bundleId'];
}
function getAdditionalBundleIds(instance) {
    return instance['_additionalBundleIds'];
}
function getSkippedApps(instance) {
    return instance['_skippedApps'];
}
function getClientEventListeners(instance) {
    return instance['_clientEventListeners'];
}
function getPageLoading(instance) {
    return instance['_pageLoading'];
}
function setPageLoading(instance, value) {
    instance['_pageLoading'] = value;
}
function getGarbageCollectOnExecute(instance) {
    return instance['_garbageCollectOnExecute'];
}
function getNavigatingToPage(instance) {
    return instance['_navigatingToPage'];
}
function setNavigatingToPage(instance, value) {
    instance['_navigatingToPage'] = value;
}
function setCurrentState(instance, value) {
    instance['_currentState'] = value;
}
function setConnectedDrivers(instance, value) {
    instance['_connectedDrivers'] = value;
}
function getPageLoadDelay(instance) {
    return instance['_pageLoadDelay'];
}
function setPageLoadDelay(instance, value) {
    instance['_pageLoadDelay'] = value;
}
function getPageLoadStartegy(instance) {
    return instance['_pageLoadStrategy'];
}
function getPageReadyTimeout(instance) {
    return instance['_pageReadyTimeout'];
}
